### ISSUE TITLE
Consider custom router ?

### ISSUE THREAD

**Issue by cuiweiqiang** [elastic/apm-agent-nodejs#878]
**Is your feature request related to a problem? Please describe.**

I'm using egg.js to develop project. In this project, I want to use apm to monitor performance which from fronted to backend service.

In this project apm can not naming transaction with pathname, because of apm use http module, but egg.js use a module forked from `koa-router`

**Describe the solution you'd like**

Egg.js fork `koa-router` and rename [egg-router](https://github.com/eggjs/egg-router), they add some additional features, but not change the property. So apm can use `koa-router` module to parse request in egg.js in theory. I notice that, it do judge with package name in `Instrumentation`. So can add a parameter like `customRouterMap`, for exmaple:
```js
  customRouterMap: {
    "@eggjs/router": {
        name: "koa-router",
        version: "7.4.0"
    }
  }
```
put it in start func, then apm can use `koa-router` module to naming transaction.

There has a simple implementation in [my fork](https://github.com/cuiweiqiang/apm-agent-nodejs/tree/customRouterMap).

Please checkout whether it is feasible .

**Comment by watson** [comment]
Interesting use-case. Thanks for spending time thinking about this coming up with a possible solution. This seems like a general solution for "when module A is being required, patch it with the code for module B" - i.e. not limited to routers only.

While this does seem like a very specific edge case, I do struggle to find an alternative. So maybe we should add support for this 🤔 

It of course comes with no guarantees as `egg-router` might change down the line in a way that means the code for patching `koa-router` no longer works on it. And without any tests in place to detect this, you're flying blind.

Let me just give it a think...

(for my own later reference, here's a link that shows the diff for just the patch: https://github.com/elastic/apm-agent-nodejs/compare/fd507f6838391467fa6076335c4c2379914d1419...cuiweiqiang:customRouterMap)

**Comment by Qard** [comment]
It could change, but even _more_ likely is the version numbers not matching, so our use of version numbers to decide whether or not to apply a given patch would no longer work. I'm not sure there is a way to do this with any reasonable degree of stability. With the patch registry incoming, I think it's better to just have users add their own patches for more obscure things. In this case, they could just do this:

```js
agent.registerPatch('@eggjs/router', require('elastic-apm-node/lib/instrumentation/modules/koa-router'))
```

**Comment by watson** [comment]
Good point about the version number @Qard - I didn't think about that.

@cuiweiqiang would you be ok with the approach @Qard described above instead? You can see more info on the registry patch he's referring to in this PR: #803

**Comment by cuiweiqiang** [comment]
@watson  @Qard  Thanks for your work. I just have one confuse problem. In Egg.js , it will start a master process and then start worker, so `apm.addPatch` will work for worker? In egg I must use apm like that:
```js
require('egg').startCluster({
  baseDir: __dirname,
  port: process.env.PORT || 7001, // default to 7001
  workers: 1,
  require: ['elastic-apm-node/start'],
});
``` 
use elastic-apm-node.js to configure apm. I'm not sure `addPatch` work or not.

**Comment by watson** [comment]
@cuiweiqiang do you mean that you need a way to specify this via config options instead of calling a function? We have that planned as well and you can set it via for instance environment variables, which I expect is what you do currently? It will look something like this:

```bash
ELASTIC_APM_PATCH=@eggjs/router=elastic-apm-node/lib/instrumentation/modules/koa-router
```

The exact name of the config option and the syntax of how you configure is not final yet, so it might change in the final release.

**Comment by cuiweiqiang** [comment]
@watson Thank you very much. This can solve lots of confuse in my project. 
👍👍

**Comment by senlin-wei** [comment]
@cuiweiqiang Excuse me, are you solved this problem, I also use egg.js, at the time of setting agent also encounter APM cannot collect data problems.  How do you to set up in index.js.And i setting config in apm-agent-node package config,and setting index.js like this

require('egg').startCluster({
  baseDir: __dirname,
  port: process.env.PORT || 7001, // default to 7001
  workers: 1,
  require: ['elastic-apm-node/start'],
});
Setting env:  ELASTIC_APM_PATCH=@eggjs/router=elastic-apm-node/lib/instrumentation/modules/koa-router

But it no effective.

**Comment by watson** [comment]
@senlin-wei no not yet, #803 is still being worked on. Once it lands, we'll update this issue with information on how to use it.

**Comment by Qard** [comment]
The patch registry has landed now and we will be releasing it soon. 😸

**Comment by watson** [comment]
The feature was just released in v2.7.0 😊

Check out the new documentation for adding custom patches here: https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#apm-add-patch

**Comment by cuiweiqiang** [comment]
@watson Good job, Thanks all yours.
